---
title: "splattDR: simulating dose-response single-cell data"
author: "Rance Nault"
package: "splattDR"
data: "Last updated: April 7, 2021"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
  vignette: >
    %\VignetteIndexEntry{simple DR simulation}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

# Installation
The development version of SplattDR can be installed from Github:
```{r}
#devtools::install(...elt())
```

# Quickstart
SplattDR begins with a `SingleCellExperiment` object. In this example we begin with a mock dataset generated using the `scater` package.  

```{r quick-start}
# Load packages
suppressPackageStartupMessages({
  library("splattdr")
  library("splatter")
  library("scater")
  library("ggplot2")
})

# Create mock data
set.seed(1)
sce = mockSCE()

# Estimate parameters from the SCE object -- replace with real dose-response
params <- splatter::splatEstimate(sce)

# Define doses and number of cells per dose
sim.dose = c(0, 0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30)
dose.prob = rep(1/9, 9)

# Simulate dose response data from parameter estimates
sim = splatSimulateDR(params, dose.names = sim.dose, dose.prob = dose.prob, verbose = FALSE)

# Visualize the simulated data
sim = runPCA(sim)
pca.df = data.frame(PC1 = reducedDim(sim)[,1], 
                    PC2 = reducedDim(sim)[,2],
                    Dose = colData(sim)$Dose
                    )

ggplot(data = pca.df, aes(x = PC1, y = PC2)) +
  geom_point(size = 1.4, aes(color = Dose)) +
  theme_bw()
```

SplattDR is a wrapper for [splatter](https://github.com/Oshlack/splatter). The following sections with outline the new components of splattDR.


# Simulating differential expression

SplattDR  uses splatter to simulate single-cell data for a _control_ group. Using baseline expression data, a set proportions of genes are randomly assigned as differentially expressed _(DE)_ using a log-normal distribution. The proportion of DE genes can be obtained from the `Params` object using `getParams'
```{r}
getParam(params, "de.prob")
```
It can also be changed using the `setParam` function:
```{r}
params = setParam(params, "de.prob", 0.5)
getParam(params, "de.prob")
```
More details on usage of the `Params` object can be obtained from the [Splatter package](https://github.com/Oshlack/splatter)

# Dose-response models

SplattDR simulates differential gene expression according to a random set of known dose-response models and records the model parameters for validation. The dose-response models, outlined in the [U.S. EPA BMDS documentation](https://www.epa.gov/bmds) are the following:
* Hill:   μ(dose)=γ+ (v dose^n)/(k^n+ dose^n )
* Exponential 2nd and 3rd:    μ(dose)=a exp⁡(sign (b*dose)^
* Exponential 3rd or 4th:     μ(dose)=a(c-(c-1)exp⁡(-1 (bdose)^d )
* Power:    μ(dose)=γ+ β dose^δ
* Linear:   μ(dose)=β_0+β_(1 ) dose+ β_(2 ) dose^2+ ⋯ + β_(n ) dose^n
* Polynomial:   μ(dose)=β_0+β_(1 ) dose+ β_(2 ) dose^2+ ⋯ + β_(n ) dose^n

The model fit outputs are saved in the `SingleCellExperiment` metadata named `modelFits`. 
```{r}
head(metadata(sim)$modelFits)
```
 As the splattDR output isa `SingleCellExperiment` object, we can obtain any cell specific metadata using `colData` and information about each gene using `rowData`. The simulated data is stores in the `assays` slots. 
```{r}
# Cell metadata
head(colData(sim))

# Gene metadaeta
head(rowData(sim))

# Log1p normalized count data
assays(sim)$logcounts[1:5, 1:5]
```

The `splatSimulateDR` function adds the following information to the splatter simulation:

* **Cell Information (`colData`)**
  * `Dose` - The dose group to which the cell is assigned
* **Gene Information (`rowData`)**
  * `DE_idx` - Differential expression factor. A value of 1 is not differentially expressed and any value not equal to 1 is used as the max |fold-change| of the dose-response model.
  * `Model` - The dose-response model used to simulate differential expression


# Other useful SplattDR functions

## Calculating real fold-change
SplattDR uses splatter to simulate gene expression by identifying a base mean value from dose-response models. It may be of value to compare the true dose-response means to the simulated means after biological noise and distribution sampling has been performed. This can be accomplished using `calcFC`.
```{r}
realFC = calcFC(sim)
head(realFC)
```

We can compare the real to simulated fold-changes
```{r}
fc.df = merge(metadata(sim)$modelFits, realFC, by = 'row.names', all.x = TRUE)
ggplot(data = fc.df, aes(x = log(fc), y  = calculatedFC30)) +
  geom_point() + 
  stat_smooth(method = "lm", col = "blue") +
  theme_bw()
```
